### PR TITLE
BUG: fix isin with nans and large arrays

### LINKED ISSUE #22205: Inconsistent handling of nan-float64 in Series.isin()
#### Code Sample, a copy-pastable example if possible

```python
N=10**6
s=pd.Series(np.full(N, np.nan, dtype=np.float64), copy=False)
s.isin([np.nan]).all()
```

results in `True`, however

```python
s=pd.Series(np.full(N+1, np.nan, dtype=np.float64), copy=False)
s.isin([np.nan]).all()
```
 results in `False`, even more  `s.isin([np.nan]).any()` results in `False`

#### Problem description

Obviously, the result should be the same in both cases.

#### Expected Output

 IMO `True` is more consistent with the behavior of Pandas in other functions (such as `pd.unique()`).

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-53-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: 3.2.1
pip: 10.0.1
setuptools: 36.5.0.post20170921
Cython: 0.28.3
numpy: 1.13.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 0.9.8
lxml: 3.8.0
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: 0.1.3
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Hanspagh**
Does a np.isnan if nan is given to isin and we have a large enough array to trigger the `np.in1d` path
- [x] closes #22205
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by Hanspagh** [comment]
Seems tests are failing, but it seems unrelated to this?

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -440,7 +440,12 @@ def isin(comps: AnyArrayLike, values: AnyArrayLike) -> np.ndarray:
     # GH16012
     # Ensure np.in1d doesn't get object types or it *may* throw an exception
     if len(comps) > 1_000_000 and not is_object_dtype(comps):
-        f = np.in1d
+        # If the the values include nan we need to check for nan explicitly
+        # since np.nan it not equal to np.nan
+        if any(np.isnan(values)):
```
Comment (dsaxton) [inline pandas/core/algorithms.py:445]:
```suggestion
        if np.isnan(values).any():
```
This would probably be faster

**Comment by dsaxton** [comment]
> Seems tests are failing, but it seems unrelated to this?

Yes, those are unrelated

**Comment by dsaxton** [comment]
Thanks @Hanspagh, can you add a release note for 1.1.3?

**Comment by Hanspagh** [comment]
Changes as requested, let mere know if there is anything more needed

File: pandas/tests/test_algos.py [hunk 2]
```diff
@@ -841,6 +840,13 @@ def test_same_nan_is_in(self):
         result = algos.isin(comps, values)
         tm.assert_numpy_array_equal(expected, result)
 
+    def test_same_nan_is_in_large(self):
```
Comment (dsaxton) [inline pandas/tests/test_algos.py:843]: It's somewhat duplicative, but could you also add a test for Series to show that the specific behavior from the bug report is fixed?

File: doc/source/whatsnew/v1.1.3.rst [hunk 3]
```diff
@@ -22,7 +22,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- Bug in :meth:`isin()` when using NaN and a row length above 1,000,000 (:issue:`22205`)
```
Comment (dsaxton) [inline doc/source/whatsnew/v1.1.3.rst:25]:
```suggestion
- Bug in :meth:`Series.isin` when using ``NaN`` and a row length above 1,000,000 (:issue:`22205`)
```

File: pandas/tests/test_algos.py [hunk 4]
```diff
@@ -841,6 +840,13 @@ def test_same_nan_is_in(self):
...
+        s = np.tile(1.0, 1_000_001)
```
Comment (jreback) [inline pandas/tests/test_algos.py:844]: pls also add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
pls also merge master, ping when green.

File: doc/source/whatsnew/v1.1.3.rst (hunk 3 above)
Comment (Hanspagh) [reply doc/source/whatsnew/v1.1.3.rst:27]: This in theory also for Dataframe.isin, I could see `isin` documented like this previously, which one is the prefered way?

**Comment by Hanspagh** [comment]
Updated as requested

File: doc/source/whatsnew/v1.1.3.rst (hunk 3 above)
Comment (dsaxton) [reply doc/source/whatsnew/v1.1.3.rst:32]: Could also mention both with :meth:\`Series.isin\` and :meth:\`DataFrame.isin\`

Comment (jreback) [reply doc/source/whatsnew/v1.1.3.rst:32]: prefer the specific ones (Series/DataFrame)

File: pandas/tests/test_algos.py [hunk 5]
```diff
@@ -841,6 +840,23 @@ def test_same_nan_is_in(self):
         result = algos.isin(comps, values)
         tm.assert_numpy_array_equal(expected, result)
 
+    # issue:`22205`
```
Comment (jreback) [inline pandas/tests/test_algos.py:843]: can you put *inside* the function

File: pandas/tests/test_algos.py [hunk 6]
```diff
@@ -841,6 +840,23 @@ def test_same_nan_is_in(self):
...
+    def test_same_nan_is_in_large(self):
+        s = np.tile(1.0, 1_000_001)
+        s[0] = np.nan
+        result = algos.isin(s, [np.nan, 1])
+        expected = np.ones(len(s), dtype=bool)
+        tm.assert_numpy_array_equal(result, expected)
+
+    # issue:`#25395`
```
Comment (jreback) [inline pandas/tests/test_algos.py:851]: same

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments, ping on green.

**Comment by Hanspagh** [comment]
Fixed

File: doc/source/whatsnew/v1.1.3.rst [hunk 7]
```diff
@@ -30,6 +30,8 @@ Bug fixes
 - Bug in :meth:`Series.str.startswith` and :meth:`Series.str.endswith` with ``category`` dtype not propagating ``na`` parameter (:issue:`36241`)
 - Bug in :class:`Series` constructor where integer overflow would occur for sufficiently large scalar inputs when an index was provided (:issue:`36291`)
 - Bug in :meth:`DataFrame.stack` raising a ``ValueError`` when stacking :class:`MultiIndex` columns based on position when the levels had duplicate names (:issue:`36353`)
+- Bug in :meth:`isin()` when using NaN and a row length above 1,000,000 (:issue:`22205`)
```
Comment (dsaxton) [inline doc/source/whatsnew/v1.1.3.rst:33]: Can you delete this note in favor of the one below?

**Comment by Hanspagh** [comment]
Done

File: pandas/tests/test_algos.py [hunk 8]
```diff
@@ -841,6 +840,23 @@ def test_same_nan_is_in(self):
         result = algos.isin(comps, values)
         tm.assert_numpy_array_equal(expected, result)
 
+    def test_same_nan_is_in_large(self):
+        # issue:`22205`
```
Comment (simonjayhawkins) [inline pandas/tests/test_algos.py:844]:
```suggestion
        # https://github.com/pandas-dev/pandas/issues/22205
```

or just `GH 22205`

File: pandas/tests/test_algos.py [hunk 9]
```diff
@@ -841,6 +840,23 @@ def test_same_nan_is_in(self):
...
+        s = np.tile(1.0, 1_000_001)
+        s[0] = np.nan
+        result = algos.isin(s, [np.nan, 1])
+        expected = np.ones(len(s), dtype=bool)
+        tm.assert_numpy_array_equal(result, expected)
+
+    def test_same_nan_is_in_large_series(self):
+        # issue:`#25395`
```
Comment (simonjayhawkins) [inline pandas/tests/test_algos.py:852]:
```suggestion
        # https://github.com/pandas-dev/pandas/issues/22205
```

or just `GH 22205`

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @Hanspagh minor nit re consistency of issue number comments

**Comment by Hanspagh** [comment]
I can do a rebase if needed?

**Comment by simonjayhawkins** [comment]
> I can do a rebase if needed?

git pull will normally get the changes from the commit suggestion locally.

to update the PR with the latest changes on master see https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#updating-your-pull-request

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Hanspagh

**Comment by simonjayhawkins** [comment]
https://github.com/pandas-dev/pandas/pull/36385#issuecomment-694275310 could have been responsible for not being able to auto backport

File: pandas/core/algorithms.py [hunk 10]
```diff
@@ -440,7 +440,12 @@ def isin(comps: AnyArrayLike, values: AnyArrayLike) -> np.ndarray:
     # GH16012
     # Ensure np.in1d doesn't get object types or it *may* throw an exception
     if len(comps) > 1_000_000 and not is_object_dtype(comps):
```
Comment (realead) [inline pandas/core/algorithms.py:442]: Sorry a little bit too late. The whole point of doing this for len(comps) > 1_000_000, was that numpy was deemed to be faster (which is probably no loner the case btw, see https://github.com/pandas-dev/pandas/issues/22205#issuecomment-410519409), adding `any`, `isnan`, `logical_or` on top  (with all the cache misses and temporary objects) will make this branch much slower. So probably it is best just to drop the whole branch and always keep `f = htable.ismember_object ` (unless it is `is_integer_dtype` of cause).

Comment (jreback) [reply pandas/core/algorithms.py:442]: can u run the asvs and check here?

Comment (realead) [reply pandas/core/algorithms.py:442]: @jreback I have opened RP #36611 with my suggestion and some benchmarks, which show that numpy's `in1d` is only faster when here are very few  unique `values`.

**Pull request merged**